Fix compiler warnings

henrygab opened the pull request:
Remove two compiler warnings.

henrygab committed:
Fix #7 - signed/unsigned mismatch

henrygab committed:
Fix #8 by init value in error path

hathach commented on Adafruit_USBD_Device.cpp:
```diff
@@ -270,7 +271,7 @@ static int strcpy_utf16(const char *s, uint16_t *buf, int bufsize)
   while (s[i] != 0) {
     uint32_t codepoint;
-    uint8_t utf8len = utf8Codepoint((const uint8_t *)s + i, &codepoint);
+    int8_t utf8len = utf8Codepoint((const uint8_t *)s + i, &codepoint);
```
this is one actually a bug, not simply a warning.

hathach commented on Adafruit_USBD_Device.cpp:
```diff
@@ -228,6 +228,7 @@ static int8_t utf8Codepoint(const uint8_t *utf8, uint32_t *codepointp)
   else if ((utf8[0] & 0xfe) == 0xfc)
     len = 6;
   else
+    *codepointp = 0
```
@henrygab you forgot the bracket, it causes a bigger issue that -1 will always be returned. I think we actually can leave it as it is and have the caller init it rather. Though I am more than OK if you can put a bracket here as well.

```C
static int strcpy_utf16(const char *s, uint16_t *buf, int bufsize)
{
  int i = 0;
  int buflen = 0;

  while (s[i] != 0) {
    uint32_t codepoint = 0; // init here is better
````

hathach reviewed (commented):
Thank you for PR as usual. There is a bit of issue with if/else without bracket. Once it is fixed, we could merge this.

henrygab committed:
stronger compliance with RFC3629

henrygab committed:
Allow BOM --> WORD JOINER conversion

henrygab commented:
Hi @hathach,

I fixed the bracket, and also fixed other issues in this code.  For example, UTF-8 does not allow encoding of code points over 0x0010_FFFF, and there are a number of octets that are always illegal in UTF-8 encoding.  I have provided links to authority for each new illegal stream.

**_NOTE:  I do not have an easy way to test this code.  Thus, please do not merge without at least basic validation._**

Please feel free to send any questions.

henrygab committed:
Clarify comment

hathach commented:
> Hi @hathach,
> 
> I fixed the bracket, and also fixed other issues in this code. For example, UTF-8 does not allow encoding of code points over 0x0010_FFFF, and there are a number of octets that are always illegal in UTF-8 encoding. I have provided links to authority for each new illegal stream.
> 
> **_NOTE: I do not have an easy way to test this code. Thus, please do not merge without at least basic validation._**
> 
> Please feel free to send any questions.

I have no idea how to test these Unicode string. @kaysievers would you mind taking a look at the changes. Else I am afraid, we will leave this PR pending for quite some time.

henrygab commented:
There is a useful table at https://github.com/cyb70289/utf8/#utf-8-coding-format.    UTF-8 decoding bugs have caused many serious issues.

I have not found a proper set of test data to validate a UTF-8 - to - codepoint decoder.

henrygab committed:
Improve detection of invalid codepoints

henrygab committed:
Tested implementation

henrygab commented:
@hathach  -- OK, I've now tested the UTF-8 decoder against **_all_** possible inputs.  I couldn't find a good test, so had to write it from scratch.  See [Gist 92a7077e](https://gist.github.com/henrygab/92a7077e4cc9b3e6b098ecaa5fa2d908).

The test tracks very little state.  It does this by relying on the fact that sorted codepoints will encode to sorted byte sequences ... this is why it can require a successful decoding to equal the prior decoded codepoint + 1.

Anyways, this should now be ready for your acceptance, as test code has now validated the correct decoding of all valid UTF-8 codepoints, and that no invalid UTF-8 will decode into a codepoint.  😁

hathach commented:
@henrygab sorry, I was on vacation, will check this out on Monday.

hathach commented on Adafruit_USBD_Device.cpp:
```diff
@@ -210,53 +210,104 @@ uint8_t const * tud_descriptor_configuration_cb(uint8_t index)
+}
+
+//
+// This function has an UNWRITTEN CONTRACT that the buffer is either:
```
too many texts :sweat_smile: :sweat_smile:

hathach reviewed (approved):
thank you very much for your great effort correcting and testing as always. Really great works so far, even writing tests yourself :+1: :+1:
